ackermann_steering_controller/src/ackermann_steering_controller.cpp

```diff
@@ -0,0 +1,520 @@
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/*
+ * Author: Bence Magyar
+ * Author: Masaru Morita
+ */
+
+#include <cmath>
+
+#include <tf/transform_datatypes.h>
+
+#include <urdf_parser/urdf_parser.h>
+
+#include <boost/assign.hpp>
+
+#include <ackermann_steering_controller/ackermann_steering_controller.h>
+
+static double euclideanOfVectors(const urdf::Vector3& vec1, const urdf::Vector3& vec2)
+{
+  return std::sqrt(std::pow(vec1.x-vec2.x,2) +
+                   std::pow(vec1.y-vec2.y,2) +
+                   std::pow(vec1.z-vec2.z,2));
+}
+
+/*
+ * \brief Check if the link is modeled as a cylinder
+ * \param link Link
+ * \return true if the link is modeled as a Cylinder; false otherwise
+ */
+static bool isCylinder(const boost::shared_ptr<const urdf::Link>& link)
+{
+  if (!link)
+  {
+    ROS_ERROR("Link == NULL.");
+    return false;
+  }
+
+  if (!link->collision)
+  {
+    ROS_ERROR_STREAM("Link " << link->name << " does not have collision description. Add collision description for link to urdf.");
+    return false;
+  }
+
+  if (!link->collision->geometry)
+  {
+    ROS_ERROR_STREAM("Link " << link->name << " does not have collision geometry description. Add collision geometry description for link to urdf.");
+    return false;
+  }
+
+  if (link->collision->geometry->type != urdf::Geometry::CYLINDER)
+  {
+    ROS_ERROR_STREAM("Link " << link->name << " does not have cylinder geometry");
+    return false;
+  }
+
+  return true;
+}
+
+/*
+ * \brief Get the wheel radius
+ * \param [in]  wheel_link   Wheel link
+ * \param [out] wheel_radius Wheel radius [m]
+ * \return true if the wheel radius was found; false other
+wise
+ */
+static bool getWheelRadius(const boost::shared_ptr<const urdf::Link>& wheel_link, double& wheel_radius)
+{
+  if (!isCylinder(wheel_link))
+  {
+    ROS_ERROR_STREAM("Wheel link " << wheel_link->name << " is NOT modeled as a cylinder!");
+    return false;
+  }
+
+  wheel_radius = (static_cast<urdf::Cylinder*>(wheel_link->collision->geometry.get()))->radius;
+  return true;
+}
+
+namespace ackermann_steering_controller{
+
+  AckermannSteeringController::AckermannSteeringController()
+    : open_loop_(false)
+    , command_struct_()
+    , wheel_separation_h_(0.0)
+    , wheel_radius_(0.0)
+    , wheel_separation_h_multiplier_(1.0)
+    , wheel_radius_multiplier_(1.0)
+    , steer_pos_multiplier_(1.0)
+    , cmd_vel_timeout_(0.5)
+    , allow_multiple_cmd_vel_publishers_(true)
+    , base_frame_id_("base_link")
+    , odom_frame_id_("odom")
+    , enable_odom_tf_(true)
+    , wheel_joints_size_(0)
+    , ns_("")
```

Comment:
Please don't use this.
controllers are suppose to read the parametres from controller_nh,

Reply:
Fix at https://github.com/ros-controls/ros_controllers/commit/be7d149661b978313a123a46c4539b80f9d4909f

Reply:
Fixed at https://github.com/ros-controls/ros_controllers/pull/356/commits/e99fd457a74658e376886cccd824b38329115f00